src/engine/TPA.cc

```diff
@@ -982,16 +983,101 @@ bool TPASplit::verifyExactPower(unsigned short power) const {
     return res == s_False;
 }
 
+vec<PTRef> TPABase::houdiniCheck(PTRef invCandidates, PTRef transition, SafetyExplanation::FixedPointType allignment) {
+    SMTConfig config;
+    MainSolver solver(logic, config, "Fixed-point checker");
+    solver.push();
+    auto candidates = topLevelConjuncts(logic, invCandidates);
+    if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+        solver.insertFormula(getNextVersion(transition));
+    } else {
+        if (allignment == SafetyExplanation::FixedPointType::LEFT) { solver.insertFormula(transition); }
+    }
+    solver.push();
+
+    // RIGHT:
+    //   rightInvariants /\ currentLevelTransition /\ getNextVersion(transition) =>
+    //     shiftOnlyNextVars(currentLevelTransition);
+    // LEFT:
+    //   leftInvariants /\ transition /\ getNextVersion(currentLevelTransition) =>
+    //     shiftOnlyNextVars(currentLevelTransition);
+    while (candidates.size() > 128) {
+        for (int i = candidates.size() - 1; i >= 1; i -= 2) {
+            PTRef n_f = logic.mkAnd(candidates[i], candidates[i - 1]);
+            candidates.pop();
+            candidates.pop();
+            candidates.push(n_f);
+            if (candidates.size() <= 128) { break; }
+        }
+    }
+    //    invCandidates.append(conjuncts);
+    //    Atr(x, x') /\ tr(x', x'') => Atr(x, x'')
+    //    or
+    //    Atr(x', x'') /\  tr(x, x') => Atr(x, x'')
+    //    Push transition once and for all
+    //    While loop externally, because we may drop smth important
+    PTRef goal = shiftOnlyNextVars(logic.mkAnd(candidates));
```

Comment:
Compute `logic.mkAnd(candidates)` only once and use the resulted `PTRef` everywhere.
No need to compute the conjunction repeatedly.

Reply:
But I need to update the formula when I update set of candidates

Reply:
That's true.
But you use the same conjunction for the goal and then when inserting into the solver. So it still makes sense to compute the conjunction only once and use it twice.